pull.go

```diff
@@ -152,76 +154,71 @@ func localImageNameForReference(ctx context.Context, store storage.Store, srcRef
 
 // Pull copies the contents of the image from somewhere else to local storage.
 func Pull(ctx context.Context, imageName string, options PullOptions) error {
-	spec := imageName
 	systemContext := getSystemContext(options.SystemContext, options.SignaturePolicyPath)
-	srcRef, err := alltransports.ParseImageName(spec)
-	if err != nil {
-		if options.Transport == "" {
-			options.Transport = util.DefaultTransport
-		}
-		logrus.Debugf("error parsing image name %q, trying with transport %q: %v", spec, options.Transport, err)
-		transport := options.Transport
-		if transport != util.DefaultTransport {
-			transport = transport + ":"
-		}
-		spec = transport + spec
-		srcRef2, err2 := alltransports.ParseImageName(spec)
-		if err2 != nil {
-			return errors.Wrapf(err2, "error parsing image name %q", imageName)
-		}
-		srcRef = srcRef2
+
+	boptions := BuilderOptions{
+		FromImage:           imageName,
+		SignaturePolicyPath: options.SignaturePolicyPath,
+		SystemContext:       systemContext,
+		PullBlobDirectory:   options.BlobDirectory,
+		ReportWriter:        options.ReportWriter,
 	}
-	if options.Quiet {
-		options.ReportWriter = nil // Turns off logging output
+
+	srcRef, transport, img, err := resolveImage(ctx, systemContext, options.Store, boptions)
+	if err != nil {
+		return err
 	}
-	var names []string
+
+	var errs *multierror.Error
 	if options.AllTags {
-		if srcRef.DockerReference() == nil {
-			return errors.New("Non-docker transport is currently not supported")
+		if transport != util.DefaultTransport {
+			return errors.New("Non-docker transport is not supported, for --all-tags pulling")
+		}
+
+		spec := transport + srcRef.DockerReference().Name()
+		srcRef, err = alltransports.ParseImageName(spec)
+		if err != nil {
+			return errors.Wrapf(err, "error getting repository tags")
 		}
 		tags, err := docker.GetRepositoryTags(ctx, systemContext, srcRef)
 		if err != nil {
 			return errors.Wrapf(err, "error getting repository tags")
 		}
 		for _, tag := range tags {
 			name := spec + ":" + tag
-			names = append(names, name)
+			if options.ReportWriter != nil {
+				options.ReportWriter.Write([]byte("Pulling " + name + "\n"))
+			}
+			ref, err := pullImage(ctx, options.Store, transport, name, options, systemContext)
```

Comment:
Non-blocking: 

For consistency with the `pullAndFindImage` caller, it would be nice if `name` here did not start with `transport`:
```go
repoName := whateverSrcRefIsRenamedTo.DockerReference().Name()
srcRef, err := alltransports.ParseImageName(transport+repoName)
// or srcRef, err := c/image/docker.NewReference(srcRef.DockerReference())
…
name = repoName + ":" + tag
```
or so.